storage/qingstor.go

```diff
@@ -0,0 +1,192 @@
+package storage
+
+import (
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/pengsrc/go-shared/convert"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	"github.com/yunify/qingstor-sdk-go/v3/config"
+	"github.com/yunify/qingstor-sdk-go/v3/request/errors"
+	"github.com/yunify/qingstor-sdk-go/v3/service"
+
+	"github.com/yunify/qsctl/constants"
+)
+
+// QingStorObjectStorage will implement ObjectStorage interface.
+type QingStorObjectStorage struct {
+	service *service.Service
+	bucket  *service.Bucket
+}
+
+// NewQingStorObjectStorage will create a new qingstor object storage.
+func NewQingStorObjectStorage() (q *QingStorObjectStorage, err error) {
+	cfg, err := config.New(
+		viper.GetString(constants.ConfigAccessKeyID),
+		viper.GetString(constants.ConfigSecretAccessKey),
+	)
+	if err != nil {
+		log.Errorf("contexts: Init config failed [%v]", err)
+		return
+	}
+
+	cfg.Host = viper.GetString(constants.ConfigHost)
+	cfg.Port = viper.GetInt(constants.ConfigPort)
+	cfg.Protocol = viper.GetString(constants.ConfigProtocol)
+	cfg.ConnectionRetries = viper.GetInt(constants.ConfigConnectionRetries)
+	cfg.LogLevel = viper.GetString(constants.ConfigLogLevel)
+
+	q = &QingStorObjectStorage{}
+
+	q.service, err = service.Init(cfg)
+	if err != nil {
+		log.Errorf("contexts: Init service failed [%v]", err)
+		return
+	}
+	return
+}
+
+// SetupBucket implements ObjectStorage.SetupBucket
+func (q *QingStorObjectStorage) SetupBucket(name, zone string) (err error) {
+	if zone == "" {
+		zone = viper.GetString(constants.ConfigZone)
+	}
+
+	if zone != "" {
+		q.bucket, _ = q.service.Bucket(name, zone)
+		return
+	}
+
+	client := &http.Client{
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+	url := fmt.Sprintf("%s://%s.%s:%d",
+		viper.GetString(constants.ConfigProtocol),
+		name,
+		viper.GetString(constants.ConfigHost),
+		viper.GetInt(constants.ConfigPort))
+
+	r, err := client.Head(url)
+	if err != nil {
+		log.Errorf("Head location failed [%v]", err)
+		return constants.ErrorExternalServiceError
+	}
+	if r.StatusCode != http.StatusTemporaryRedirect {
+		log.Infof("Detect bucket location failed, please check your input")
```

Comment:
In this case, should we use info or warning level? Just ask, cuz I'm not sure.

Reply:
I'm not so sure either, we need a log rules for qsctl later. For now, let's just keep them.